llvm/lib/Target/AIE/AIEFinalizeBundle.cpp

```diff
@@ -20,8 +20,17 @@ using namespace llvm;
 namespace {
 
 bool isBundleCandidate(MachineBasicBlock::instr_iterator MII) {
+  const MachineBasicBlock *MBB = MII->getParent();
+  auto *MF = MBB->getParent();
+  auto &Subtarget = MF->getSubtarget();
+  const AIEBaseInstrInfo *TII =
+      static_cast<const AIEBaseInstrInfo *>(Subtarget.getInstrInfo());
   MachineInstr *MI = &*MII;
-  if (MI->isMetaInstruction() || MII->isBundled())
+  // LoopStart should be lowered before reaching here
+  assert(!TII->isHardwareLoopStart(MI->getOpcode()) &&
+         "LoopStart is not lowered \n\r");
```

Comment:
Hi @martien-de-jong,

I can trigger this assert with the following code:

```
target datalayout = "e-m:e-p:20:32-i1:8:32-i8:8:32-i16:16:32-i32:32:32-f32:32:32-i64:32-f64:32-a:0:32-n32"
target triple = "aie2"

define i32 @main(i1 %cmp) {
entry:
  %call = tail call i32 null(ptr null, ptr null)
  br i1 %cmp, label %if.else, label %for.cond.preheader

for.cond.preheader:                          ; preds = %entry
  store i8 0, ptr null, align 4
  br label %for.body

if.else:                                      ; preds = %entry
  %bf = load i16, ptr null, align 4
  br label %for.body

exit:                        ; preds = %for.body
  ret i32 0

for.body:                                  ; preds = %for.body, %if.else, %for.cond.preheader
  %j = phi i32 [ %inc, %for.body ], [ 0, %if.else ], [ 0, %for.cond.preheader ]
  %0 = trunc i32 %j to i20
  %inc = add i32 %j, 1
  %exitcond = icmp eq i32 %inc, 0
  br i1 %exitcond, label %exit, label %for.body
}

```

Reply:
I'm on it. Probably an assumption on seeing LoopStart before LoopEnd.

Reply:
The true story: tail duplication copied LoopStart out into predecessor blocks. This scares hwloop expansion.